flash: fix `flash_read` API for NXP LPC55S36

pillo79 opened the pull request:
This PR reworks the LPC55S36 implementation of the `flash_mcux_read()` function to work around issue #80325. Feedback welcome!

This is achieved on the `CONFIG_CHECK_BEFORE_READING` path (the default) by using the `FLASH_Read()` HAL API call, which is resilient to errors, and map the error codes to properly emulate an erased page read. 

When `CONFIG_CHECK_BEFORE_READING` is disabled, the code reverts to the previous `memcpy()` operation, which is faster but unsafe on erased or otherwise unreadable pages.

`#ifdef` guards kept in the style of the file, though some may be changed to `if (IS_ENABLED(...))` if desired.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/80325

pillo79 committed:
drivers/flash/mcux: fix flash_read() operation on LPC55S36

As other targets in the LPC55xxx series, the LPC55S36 has a Flash
controller that raises ECC errors when reading erased pages directly.
To avoid this, there is special code for this platform that calls the
HAL FLASH_IsFlashAreaReadable() function. However, this in turn calls a
function at an hardcoded address in ROM that _always_ causes an
instruction fault, making the situation worse.

This patch reworks the read operation to use the FLASH_Read() HAL
function for this target to gracefully handle error conditions and
properly emulate accesses to erased pages. The preprocessor is required
since some targets do not define the FLASH_Read() function.

Fixes: #80325

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 commented:
Friendly ping for any feedback after a week of silence :slightly_smiling_face:  
I know Zephyr is in freeze right now, but since this is a bugfix I think this would be useful in 4.0.

danieldegrasse commented on drivers/flash/soc_flash_mcux.c:
```diff
@@ -204,36 +204,49 @@ static int flash_mcux_read(const struct device *dev, off_t offset,
-	} else {
-		/* Check whether the flash is erased ("len" and "addr" must be word-aligned). */
-		rc = FLASH_VerifyErase(&priv->config, ((addr + 0x3) & ~0x3),  ((len + 0x3) & ~0x3));
+#if defined(CONFIG_CHECK_BEFORE_READING) && !defined(CONFIG_SOC_LPC55S36)
```
A suggestion here for readability- could we do something like this:

```
if (IS_ENABLED(CONFIG_CHECK_BEFORE_READING)) {
    if (IS_ENABLED(CONFIG_SOC_LPC55S36) {
        /* LPC55S36 specific logic */
    } else {
       rc = is_area_readable(addr, len)
    }
}

if (!rc) {
   memcpy(data, (void *)addr, len);
} else if ((rc == -ENODATA) && IS_ENABLED(CONFIG_CHECK_BEFORE_READING)) {
  /* Erased area, return dummy data as an erased page. */
  memset(data, 0xFF, len);
  rc = 0;
}
return rc;
```

Apologies if formatting there is confusing. At least IMO, the current function implementation in `main` is hard to follow, and I figure if we are fixing the bug cleaning up the readability wouldn't hurt either

pillo79 replied:
Completely agree, but it's really a thorny situation. The `FLASH_Read` does a "copy before verify" which doesn't neatly fit in the previous model:

| `CHECK` is set? | on `55S36`? | Check stage | Copy stage | `0xff` fill 
|--------|--------|--------|-------|-------|
| no | - | no | `memcpy` | no |
| yes| no | `is_area_readable` | `memcpy` | yes |
| yes| yes| `FLASH_Read` + `VerifyErased` | (already done) | yes

My first proposal avoided any duplicate code, but ended up creating awkward logical groupings.

danieldegrasse replied:
Thank you for the explanation here, this makes more sense (and I really like the rework in the latest push, it is a lot easier to follow)

pillo79 commented:
**v2**:
* converted `#ifdef ...` to `if (IS_ENABLED(...))` 
* spelled out all cases to make it way easier to read

danieldegrasse commented on drivers/flash/soc_flash_mcux.c:
```diff
@@ -204,36 +207,54 @@ static int flash_mcux_read(const struct device *dev, off_t offset,
+		 * Ensure the area is readable, since a direct access may cause faults on erased
+		 * or otherwise unreadable pages. Emulate erased pages, return other errors.
+		 */
+		if (IS_ENABLED(CONFIG_SOC_LPC55S36)) {
```
@pillo79 it looks like we may need more "#ifdefery", because some older NXP platforms don't define the `FLASH_Read` API at all. I think we probably need to use `#ifdef` here. Sorry for the poor guidance

danieldegrasse commented:
On the next push can you reference the github issue within the commit message? IE add

`Fixes #80325` to the commit message (at the end, before the signed-off-by-line)

pillo79 committed:
drivers/flash/soc_flash_mcux: adjust alignment logic

The current alignment logic does not work as expected if given unaligned
values, resulting in a skip of the first word. The length also has to
take into account the starting address: for example, asking for 2 bytes
at offset 3 should actually check 8 bytes.

This patch adjusts the logic so that it always includes the first and
the last word of the input area.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 commented:
Minor addition I spotted only now - fixed the alignment logic for `FLASH_VerifyErase`, which was too simple and did not properly enlarge the area.

danieldegrasse commented:
@de-nordic can you take a look?

de-nordic commented on drivers/flash/soc_flash_mcux.c:
```diff
@@ -205,35 +205,55 @@ static int flash_mcux_read(const struct device *dev, off_t offset,
+		/* Check id the ECC issue is due to the Flash being erased
+		 * ("addr" and "len" must be word-aligned for this call).
+		 */
+		rc = FLASH_VerifyErase(&priv->config,
```
Question: wouldn't it be better to just ifdef the `is_area_readable` internally to use this method of verification, whether memory is readable, for `CONFIG_SOC_LPC55S36`, and then follow it with the memcopy or memset depending on result? Instead of reading first and then checking why it failed, and have streamlined code fall all devices supported?

As far as I understand the description for `FLASH_VerifyErase` in fsl_iap.h, it is able to tell whether there is ECC problem in range (btw NXP, the function no longer takes margin as parameter, yet you still mention it in description).

Additional note: IMHO making read check configurable makes no sense, because there is always a chance that ECC error will topple device instead of returning with error, unless of course you disable ECC,

de-nordic reviewed (approved):
Looks OK, as far as I can tell.
Left a comment on potential place for streamlining execution here https://github.com/zephyrproject-rtos/zephyr/pull/80351/files#r1831641313, but that is up to the vendor here.
